### PR TITLE
feat: pragma randomness

### LINKED ISSUE #554: Building Advanced Contracts: randomness section
Create a subsection in "Building Advanced Contracts', providing a contract example requesting a random felt252 using pragma VRF  for some raffle game.

### PR DESCRIPTION
**Pull request by okhaimie-dev**
Closes #554 

- - -
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/721)

### CODE REVIEW

**Comment by TAdev0** [comment]
@okhaimie-dev  great thanks! will review later today

File: src/ch16-05-02-randomness.md [hunk 1]
```diff
@@ -1 +1,58 @@
...
+Listing {{#ref pragma_vrf_interface}} shows a simple interface for an example randomness contract that uses Pragma VRF:
+
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:randomness_interface}}
+```
+
+{{#label pragma_vrf_interface}}
+<span class="caption">Listing {{#ref pragma_vrf_interface}}: Simple interface for using Pragma VRF</span>
+
+```rust,noplayground
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:27]: place the code after the text below

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        // AccessControl-related initialization
+        self.accesscontrol.initializer();
+        self.accesscontrol._grant_role(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    #[abi(embed_v0)]
+    impl Randomness of IRandomness<ContractState> {
+        fn get_last_random(self: @ContractState) -> felt252 {
+            let last_random = self.last_random_storage.read();
+            return last_random;
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:94]: instead of `return last_random;`, you can skip a line and simple use the expression `last_random` with semicolon

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    );
+    fn withdraw_funds(ref self: TContractState, receiver: ContractAddress);
+}
+//ANCHOR_END: randomness_interface
+
+//ANCHOR: randomness_contract
+#[starknet::contract]
+mod Randomness {
+    use starknet::{ContractAddress};
+    use starknet::info::{get_block_number, get_caller_address, get_contract_address};
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:35]: this doesn't compile with edition 2023_11 because of visibility issue

The right path is :

    `use starknet::{get_block_number, get_caller_address, get_contract_address};`

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            let randomness_dispatcher = IRandomnessDispatcher {
+                contract_address: randomness_contract_address
+            };
+            let caller = get_caller_address();
+            let _compute_fees = randomness_dispatcher.compute_premium_fee(caller);
+
+            // Approve the randomness contract to transfer the callback fee
+            // You would need to send some ETH to this contract first to cover the fees
+            let eth_dispatcher = ERC20ABIDispatcher {
+                contract_address: 0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7 // ETH Contract Address
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:117]: you can use : 

```
starknet::contract_address_const::<0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7>()
```

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                    .try_into()
+                    .unwrap()
+            };
+            eth_dispatcher
+                .approve(
+                    randomness_contract_address,
+                    (callback_fee_limit + callback_fee_limit / 5).into()
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:124]: does this  come from Pragma?

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            // Request the randomness
+            let _request_id = randomness_dispatcher
+                .request_random(
+                    seed, callback_address, callback_fee_limit, publish_delay, num_words, calldata
+                );
+
+            let current_block_number = get_block_number();
+            self.min_block_number_storage.write(current_block_number + publish_delay);
+
+            return ();
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:136]: remove this line, it is not necessary to explicitly return the unit type, it is returned by default when there is no return value

File: src/ch16-05-02-randomness.md [hunk 7]
```diff
@@ -1 +1,58 @@
...
+
+The function `request_my_randomness` initiates a request for verifiable randomness from the Pragma Oracle. It does this by emitting an event that triggers the following actions off-chain:
+
+1. Randomness Generation: The Oracle generates random values and a corresponding proof.
+2. On-chain Submission: The Oracle submits the generated randomness and proof back to the blockchain via the `receive_random_words` callback function.
+
+### `request_my_randomness` Inputs
+
+1. `seed`: A value used to initialize the randomness generation process. This should be unique for each request to ensure unpredictable results.
+2. `callback_address`: The contract address where the receive_random_words function will be called to deliver the generated randomness. It is typically the address of your deployed contract implementing pragma vrf.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:39]: `receive_random_words` with `

File: src/ch16-05-02-randomness.md [hunk 8]
```diff
@@ -1 +1,58 @@
...
+3. `callback_fee_limit`: The maximum amount of gas you're willing to spend on executing the receive_random_words callback function.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:40]: same

File: src/ch16-05-02-randomness.md [hunk 9]
```diff
@@ -1 +1,58 @@
...
+4. `publish_delay`: The minimum delay (in blocks) between requesting randomness and the Oracle fulfilling the request. This allows time for the Oracle to generate the randomness and proof.
+5. `num_words`: The number of random values (each represented as a felt252) you want to receive in a single callback.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:42]: same for felt252, use `

File: src/ch16-05-02-randomness.md [hunk 10]
```diff
@@ -1 +1,58 @@
...
+6. `calldata`: Additional data you want to pass to the receive_random_words callback function. This could be used for context or to customize the behavior of that function.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:43]: same

File: src/ch16-05-02-randomness.md [hunk 11]
```diff
@@ -1 +1,58 @@
...
+
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:44]: Explain shortly other functions of the interface before sending the whole contract

File: src/ch16-05-02-randomness.md [hunk 12]
```diff
@@ -1 +1,58 @@
...
+Listing {{#ref pragma_vrf_contract}} shows a basic implementation of a randomness contract using Pragma VRF:
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:45]: of a randomness contract implementing the previous interface, and remove ''using pragma vrf''

File: src/ch16-05-02-randomness.md [hunk 13]
```diff
@@ -1 +1,58 @@
...
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:randomness_contract}}
+```
+
+{{#label pragma_vrf_contract}}
+<span class="caption">Listing {{#ref pragma_vrf_contract}}: Simple Randomness Contract</span>
+
+### NB: After Contract is Deployed
+
+Once contract is deployed, you need to send enough ETH to the deployed contract address required to cover gas cost of callback function and cost of generating randomness.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:56]: Suggestion:

Once your contract is deployed, it needs to hold enough ETH in order to use Pragma VRF system. Indeed,  requesting random values implies to pay the cost forthe random generation process and the callback function. 

I think my suggestion is not perfect but this is just an idea

File: src/ch16-05-02-randomness.md [hunk 14]
```diff
@@ -1 +1,58 @@
...
+
+For more information, refer to the [Pragma](https://docs.pragma.build/Resources/Cairo%201/randomness/randomness) docs.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:58]: please refer*

**Review by TAdev0 — CHANGES_REQUESTED** [review]

File: src/ch16-05-02-randomness.md (hunk 10 above)
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:43]: what do you mean by same here?

File: src/ch16-05-02-randomness.md (hunk 11 above)
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:44]: alright, I'll do this

File: src/ch16-05-02-randomness.md (hunk 13 above)
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:56]: alright. i'll edit and improve on it

File: src/ch16-05-02-randomness.md [hunk 15]
```diff
@@ -1 +1,58 @@
 # Randomness
+
+If we need to generate randomness on the blockchain, it has to be done in a secure manner to ensure that the result cannot be predicted or manipulated. A common way to achieve this is to use verifiable random function ( VRF ) which is typically provided by an oracle.
```
[the remark this answers is not in the document]
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:3]: Reading now. I'll make an update in a bit

File: src/ch16-05-02-randomness.md (hunk 11 above)
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:70]: So I didn't explain the
`get_last_random` function because it is just a getter function to read the `last_random` generated.
`withdraw_funds` function because it is an optional function used by contract admin to withdraw the funds sent to be used by the `receive_random_words` callback function

if you still want it in the book, i will go ahead and add these two functions

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo (hunk 5 above)
Comment (okhaimie-dev) [reply listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo]: this is a clearer version of the expression `(callback_fee_limit + (callback_fee_limit / 5)).into()`

and yes, i found out that this is a safe value that works. just the `callback_fee_limit` alone returns an error

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+    fn request_my_randomness(
+        ref self: TContractState,
+        seed: u64,
+        callback_address: ContractAddress,
+        callback_fee_limit: u128,
+        publish_delay: u64,
+        num_words: u64,
+        calldata: Array<felt252>
+    );
+
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:19]: here just remove empty line 19 and its good

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo (hunk 4 above)
Comment (TAdev0) [reply listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:123]: @okhaimie-dev  instead of doing try_into and unwrap to convert a felt252 in a contractAddress type, use `contract_address_const` from corelib, i guess it saves steps

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+            // You would need to send some ETH to this contract first to cover the fees
+            let eth_dispatcher = ERC20ABIDispatcher {
+                contract_address: 0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7 // ETH Contract Address
+                    .try_into()
+                    .unwrap()
+            };
+            eth_dispatcher
+                .approve(
+                    randomness_contract_address,
+                    (callback_fee_limit + (callback_fee_limit / 5)).into()
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:130]: you dont need the parenthesis as the division will be executed before the addition anyway

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+
+            // and that the requester_address is what we expect it to be (can be self
+            // or another contract address), checking for self in this case
+            //let contract_address = get_contract_address();
+            //assert(requester_address == contract_address, 'requester is not self');
+
+            // Optionally: Can also make sure that request_id is what you expect it to be,
+            // and that random_words_len==num_words
+
+            // Your code using randomness!
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:170]: you can remove this line i think

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+            let random_word = *random_words.at(0);
+
+            self.last_random_storage.write(random_word);
+        }
+
+        fn withdraw_funds(ref self: ContractState, receiver: ContractAddress) {
+            self.accesscontrol.assert_only_role(DEFAULT_ADMIN_ROLE);
+            let eth_dispatcher = ERC20ABIDispatcher {
+                contract_address: 0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7 // ETH Contract Address
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:179]: same, use `contract_address_const`

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+                    .try_into()
+                    .unwrap()
+            };
+            let balance = eth_dispatcher.balance_of(get_contract_address());
+            eth_dispatcher.transfer(receiver, balance);
+        }
+    }
+}
+//ANCHOR_END: randomness_contract
+
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:189]: remove empty line 189 so that there is no extra empty line in the global file

File: src/ch16-05-02-randomness.md [hunk 21]
```diff
@@ -1 +1,84 @@
 # Randomness
+
+Since most well-know blockchains are fundamentally deterministic and public ledgers, generating truly unpredictatable randomness on-chain presents a challenge. This randomness is crucial for fair outcomes in gaming, lotteries, and unique generation of NFTs. To address this, verifiable random functions (VRFs) provided by oracles offers a solution. VRFs guarantee that the randomness can't be predicted or tampered with, ensuring trust and transparency in these applications.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:3]: well-known

Comment (TAdev0) [reply src/ch16-05-02-randomness.md:3]: offer*

File: src/ch16-05-02-randomness.md [hunk 22]
```diff
@@ -1 +1,84 @@
...
+
+## Overview on VRFs
+
+Pseudo-random but secure: VRFs use a secret key and a nonce (a unique input) to generate an output that appears random. While technically 'pseudo-random', it's practically impossible for another party to predict the outcome without knowing the secret key.
+
+Verifiable output: VRFs produce not only the random number but also a proof that anyone can use to independently verify that the result was generated correctly according to the function's parameters.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:9]: nice

File: src/ch16-05-02-randomness.md [hunk 23]
```diff
@@ -1 +1,84 @@
...
+
+[Pragma](https://www.pragma.build/), an oracle on Starknet provides a solution for generating random numbers using VRFs.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:11]: colon after ''Starknet''

File: src/ch16-05-02-randomness.md [hunk 24]
```diff
@@ -1 +1,84 @@
...
+```
+
+Listing {{#ref pragma_vrf_interface}} shows a contract interface of an example randomness contract that uses Pragma VRF:
+
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:randomness_interface}}
+```
+
+{{#label pragma_vrf_interface}}
+<span class="caption">Listing {{#ref pragma_vrf_interface}}: Simple Randomness Contract Interface</span>
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:31]: add a point after ''Interface''

File: src/ch16-05-02-randomness.md [hunk 25]
```diff
@@ -1 +1,84 @@
...
+
+The function `request_my_randomness` initiates a request for verifiable randomness from the Pragma Oracle. It does this by emitting an event that triggers the following actions off-chain:
+
+1. Randomness Generation: The Oracle generates random values and a corresponding proof.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:35]: the oracle without capital letter

File: src/ch16-05-02-randomness.md [hunk 26]
```diff
@@ -1 +1,84 @@
...
+2. On-chain Submission: The Oracle submits the generated randomness and proof back to the blockchain via the `receive_random_words` callback function.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:36]: same

File: src/ch16-05-02-randomness.md [hunk 27]
```diff
@@ -1 +1,84 @@
...
+
+
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:38]: remove empty line in excess

File: src/ch16-05-02-randomness.md [hunk 28]
```diff
@@ -1 +1,84 @@
...
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:request_my_randomness}}
+```
+
+<span class="caption">Listing {{#ref request_my_randomness}}: `request_my_randomness` function</span>
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:43]: point after ''function''

File: src/ch16-05-02-randomness.md [hunk 29]
```diff
@@ -1 +1,84 @@
...
+5. `num_words`: The number of random values (each represented as a `felt252`) you want to receive in a single callback.
+6. `calldata`: Additional data you want to pass to the `receive_random_words` callback function.
+
+The function `receive_random_words` is a callback triggered by the Pragma Randomness Oracle when it has generated the randomness requested by your contract.
+
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:receive_random_words}}
+```
+
+<span class="caption">Listing {{#ref receive_random_words}}: `receive_random_word` function</span>
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:60]: same, point after ''function''

File: src/ch16-05-02-randomness.md (hunk 11 above)
Comment (TAdev0) [reply src/ch16-05-02-randomness.md:70]: @enitrat  what do you think?

File: src/ch16-05-02-randomness.md [hunk 30]
```diff
@@ -1 +1,84 @@
...
+```rust,noplayground
+{{#include ../listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:randomness_contract}}
+```
+
+{{#label pragma_vrf_contract}}
+<span class="caption">Listing {{#ref pragma_vrf_contract}}: Simple Randomness Contract</span>
+
+### NB: After Contract is Deployed
+
+After deploying your contract, ensure it holds sufficient ETH to use the Pragma VRF system. Requesting random values incurs costs associated with both the generation process and the execution of your callback function.
```
Comment (TAdev0) [inline src/ch16-05-02-randomness.md:82]: nice

**Review by TAdev0 — CHANGES_REQUESTED** [review]

File: src/ch16-05-02-randomness.md (hunk 21 above)
Comment (okhaimie-dev) [reply src/ch16-05-02-randomness.md:3]: i took out the "well-know". 

> Since most blockchains

still conveys the message

Comment (TAdev0) [reply src/ch16-05-02-randomness.md:3]: Actually all blockchain systems or any computer is deterministic , and most of them are public ledgers

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo (hunk 20 above)
Comment (okhaimie-dev) [reply listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:179]: done

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 31]
```diff
@@ -0,0 +1,182 @@
...
+                >() // ETH Contract Address
+            };
+            eth_dispatcher
+                .approve(
+                    randomness_contract_address,
+                    (callback_fee_limit + callback_fee_limit / 5).into()
+                );
+
+            // Request the randomness
+            let _request_id = randomness_dispatcher
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:134]: Unused local variable ?

File: listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+            publish_delay: u64,
+            num_words: u64,
+            calldata: Array<felt252>
+        ) {
+            let randomness_contract_address = self.randomness_contract_address.read();
+            let randomness_dispatcher = IRandomnessDispatcher {
+                contract_address: randomness_contract_address
+            };
+            let caller = get_caller_address();
+            let _compute_fees = randomness_dispatcher.compute_premium_fee(caller);
```
Comment (TAdev0) [inline listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo:118]: Remove underscore before compute

**Review by TAdev0 — CHANGES_REQUESTED** [review]

**Review by enitrat — CHANGES_REQUESTED** [review]
I would like to see a more specific usage of randomness in an application rather than a contract that doesn't have a real purpose. 

How about implementing a simple ERC20-faucet dice game where the user can call an entrypoint with a number between 1 and 6, we store the user's input and request randomness, and when we receive a response in the randomness callback, we distribute 1 token to each user that guessed the right number ?

That would be a better guide on how to integrate randomness with a raffle game. 

> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/721#-:-Nv2XQNzCQk7aHzWEokH:b-cs5ev5)* status: 1 of 5 files reviewed, 45 unresolved discussions (waiting on @okhaimie-dev and @TAdev0)

___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 94 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/721#gh-1554854721:-Nv2VEhv-Kq7Pya5-xkX:bsluw1g) ([raw file](https://github.com/cairo-book/cairo-book/blob/b583a4a4adaba81e7ca38e1597e0f168f4f6f6dd/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L94)):*
<details><summary><i>Previously, TAdev0 (Tristan) wrote…</i></summary><blockquote>

instead of `return last_random;`, you can skip a line and simple use the expression `last_random` with semicolon
</blockquote></details>

simply `self.last_random_storage.read()` 
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 124 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/721#gh-1554856626:-Nv2ViPKA0qcpDV1Y4fz:bkf895d) ([raw file](https://github.com/cairo-book/cairo-book/blob/b583a4a4adaba81e7ca38e1597e0f168f4f6f6dd/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L124)):*
<details><summary><i>Previously, okhaimie-dev (okhai.stark ( Tony Stark )) wrote…</i></summary><blockquote>

this is a clearer version of the expression `(callback_fee_limit + (callback_fee_limit / 5)).into()`

and yes, i found out that this in one of their example code. i also found it to be a safe value for approval
</blockquote></details>

it comes a bit out of nowhere, it's hard to follow along
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 34 at r6](https://reviewable.io/reviews/cairo-book/cairo-book/721#-Nv2VufKEWqpX00BI-TS:-Nv2VufLCxDI78KElvyn:bibj1nw) ([raw file](https://github.com/cairo-book/cairo-book/blob/27051648a2042bba33b46d002774d4c177be560e/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L34)):*
> ```Cairo
> }
> //ANCHOR_END: randomness_interface
> ```

Where is the interface of the Pragma contract? Is it the same as the IRandomness contract you're implementing? it's not very clear what we're requesting random numbers for
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 37 at r6](https://reviewable.io/reviews/cairo-book/cairo-book/721#-Nv2Uygv68tzhXph7Mr8:-Nv2Uygv68tzhXph7Mr9:bdkfmct) ([raw file](https://github.com/cairo-book/cairo-book/blob/27051648a2042bba33b46d002774d4c177be560e/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L37)):*
> ```Cairo
> //ANCHOR: randomness_contract
> #[starknet::contract]
> mod Randomness {
> ```

It is not clear what this contract actually does from the name. I see it's called `Randomness`-  but it somehow interacts with tokens, so it must have an applicative value? The reader should know what the contract does by reading its name and the functions in the interface.
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 61 at r6](https://reviewable.io/reviews/cairo-book/cairo-book/721#-Nv2UmUP09-Web0fOikZ:-Nv2UmUP09-Web0fOik_:brmfcya) ([raw file](https://github.com/cairo-book/cairo-book/blob/27051648a2042bba33b46d002774d4c177be560e/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L61)):*
> ```Cairo
>     // SRC5
>     #[abi(embed_v0)]
>     impl SRC5Impl = SRC5Component::SRC5Impl<ContractState>;
> ```

What is this for? I don't think it is needed
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 113 at r6](https://reviewable.io/reviews/cairo-book/cairo-book/721#-Nv2VJRT9UTBRVYJXa2w:-Nv2VJRT9UTBRVYJXa2x:b-sf7yq8) ([raw file](https://github.com/cairo-book/cairo-book/blob/27051648a2042bba33b46d002774d4c177be560e/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L113)):*
> ```Cairo
>             calldata: Array<felt252>
>         ) {
>             let randomness_contract_address = self.randomness_contract_address.read();
> ```

is this Pragma VRF? It should be named as such. It's confusing, I don't know what this refers to when I read this contract. Your contract is also named randomness
___
*[`listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo` line 114 at r6](https://reviewable.io/reviews/cairo-book/cairo-book/721#-Nv2VQhEC2cJtmsCI1Vz:-Nv2VQhEC2cJtmsCI1W-:bgnt9im) ([raw file](https://github.com/cairo-book/cairo-book/blob/27051648a2042bba33b46d002774d4c177be560e/listings/ch16-building-advanced-starknet-smart-contracts/listing_06_pragma_randomness/src/lib.cairo#L114)):*
> ```Cairo
>         ) {
>             let randomness_contract_address = self.randomness_contract_address.read();
>             let randomness_dispatcher = IRandomnessDispatcher {
> ```

can be merged with the line above
___
*[`src/ch16-05-02-randomness.md` line 14 at r5](https://reviewable.io/reviews/cairo-book/cairo-book/721#gh-1556423958:-Nv2WMGP6jebHrw01qsd:b-mftzmj) ([raw file](https://github.com/cairo-book/cairo-book/blob/c43ab96c1cab85cf3579af5423a59238336fe049/src/ch16-05-02-randomness.md#L14)):*
<details><summary><i>Previously, TAdev0 (Tristan) wrote…</i></summary><blockquote>

I think we should modify the flow for titles, maybe something like this? :
```
## Overview on VRFs

## Example of a Randomness Contract (maybe not ideal though)
precise here that our example will use Pragma, the most used oracle on starknet
and then 

### Add Pragma as a Dependency

### Interface 

### Description of Entrypoints and their inputs

### Contract Example -  (full implementation)

Because currently, the flow is:
## Overview on VRFs
## Add Pragma as a Dependency
### `request_my_randomness` Inputs
### `receive_randomn_words` Inputs
## Example Randomness Contract
### NB: After Contract is Deployed

and its not logic 

```
</blockquote></details>

there is an issue with the current flow of the chapter: it should be implemented as a follow-along implementation of writing a contract that uses Pragma VRF, but in the current state it's hard to actually understand what's going on.

@TAdev0 even your proposal remains purely descriptive and not informative enough. 

When you write a contract, you
- Describe your use case and application
- Think about what will be exposed to the outside world, and how (view / external)

Then you start writing the code
- Create your project with the right dependencies and file structure
- Write the interface of your contract
- Implement your interface, feature by feature

Step by step, gradually.

**Comment by okhaimie-dev** [comment]
> is this Pragma VRF? It should be named as such. It's confusing, I don't know what this refers to when I read this contract. Your contract is also named randomness

it is the pragma randomness contract address. I will make the variable names more descriptive in my next commit using the new example you suggested

**Comment by enitrat** [comment]
Hey @okhaimie-dev are you going to revise this? Do you need any help?

**Comment by okhaimie-dev** [comment]
@enitrat 

I'll complete it this week. I'll reach out if I need further assistance MAX by Wednesday

**Comment by okhaimie-dev** [comment]
I am still working on the Dice game example.
https://github.com/okhaimie-dev/dice-game

I will used this to replace the initial example I've got in a bit

**Comment by enitrat** [comment]
Closing this, you can reopen a new one when you're ready
